aws-api/src/main/java/com/amplifyframework/api/aws/SubscriptionEndpoint.java

```diff
@@ -305,34 +335,42 @@ private void notifySubscriptionData(String subscriptionId, String data) throws A
     }
 
     synchronized void releaseSubscription(String subscriptionId) throws ApiException {
-        final Subscription<?> subscription = subscriptions.get(subscriptionId);
-        if (subscription == null) {
+        // First thing we should do is remove it from the subscriptions collections so
+        // the other methods can't grab a hold of the subscription.
+        final Subscription<?> subscription = subscriptions.remove(subscriptionId);
+        boolean wasSubscriptionPending = pendingSubscriptions.remove(subscriptionId);
+        // If the subscription was not in the subscriptions collections AND was also
+        // not in the pending collection.
+        if (subscription == null && !wasSubscriptionPending) {
             throw new ApiException(
                 "No existing subscription with the given id.",
                 AmplifyException.TODO_RECOVERY_SUGGESTION
             );
         }
 
-        try {
-            webSocket.send(new JSONObject()
-                .put("type", "stop")
-                .put("id", subscriptionId)
-                .toString());
-        } catch (JSONException jsonException) {
-            throw new ApiException(
-                "Failed to construct subscription release message.",
-                jsonException,
-                AmplifyException.TODO_RECOVERY_SUGGESTION
-            );
+        // Only do this if the subscription was NOT pending.
+        // Otherwise it would probably fail since it was never established in the first place.
+        if (!wasSubscriptionPending) {
+            try {
+                webSocket.send(new JSONObject()
+                    .put("type", "stop")
+                    .put("id", subscriptionId)
+                    .toString());
+            } catch (JSONException jsonException) {
+                throw new ApiException(
+                    "Failed to construct subscription release message.",
+                    jsonException,
+                    AmplifyException.TODO_RECOVERY_SUGGESTION
+                );
+            }
+            subscription.awaitSubscriptionCompleted();
         }
 
-        subscription.awaitSubscriptionCompleted();
-        subscriptions.remove(subscriptionId);
-
         // If we have zero subscriptions, close the WebSocket
         if (subscriptions.size() == 0) {
             timeoutWatchdog.stop();
             webSocket.close(NORMAL_CLOSURE_STATUS, "No active subscriptions");
+            webSocket = null; //force creation of a new websocket
```

Comment:
Ahh. Nice catch.

Reply:
It ended up looking a little different, but the spirit is the same. Basically need to force the creation of a new websocket.